add support for including commit pattern

rouzwelt opened the pull request:
proposing to add support for including commits pattern for changelog. with the option to make it strict with "--strict-include" to whether include merge commits or not.

if the strict flag is provided then only include based on provided regex, else include merge commits as well.

ljharb commented on src/commits.js:
```diff
@@ -35,7 +35,8 @@ const parseCommits = (string, options = {}) => {
     .slice(1)
     .map(commit => parseCommit(commit, options))
-    .filter(commit => filterCommit(commit, options))
+    .filter(commit => blacklistCommits(commit, options))
```
and here, "exclude list" perhaps?

rouzwelt replied:
hey @ljharb,
yes sure, I'll update it

ljharb reviewed (changes requested):
Please use a term other than "whitelist" - perhaps "allowlist"?

ljharb commented on README.md:
```diff
@@ -21,40 +21,42 @@ Usage: auto-changelog [options]
+      --breaking-pattern [regex]        # regex pattern for breaking change commits
+      --merge-pattern [regex]           # add custom regex pattern for merge commits
+      --exclude-commit-pattern [regex]  # pattern to exclude when parsing commits
+      --include-commit-pattern [regex]  # pattern to include when parsing commits, merge commits are be included by default unless "--strict-include" is chekced
```
```suggestion
      --include-commit-pattern [regex]  # pattern to include when parsing commits, merge commits are be included by default unless "--strict-include" is checked
```

ljharb commented on src/commits.js:
```diff
@@ -35,7 +35,8 @@ const parseCommits = (string, options = {}) => {
     .slice(1)
     .map(commit => parseCommit(commit, options))
-    .filter(commit => filterCommit(commit, options))
+    .filter(commit => excludeCommit(commit, options))
+    .filter(commit => includeCommit(commit, options))
```
what happens if include and exclude have overlaps? that seems like it should be an error rather than just arbitrarily deciding which one takes precedence.

rouzwelt replied:
well I thought about this for while before making this PR, and what I concluded after a bit of tests was that it doesn't really matter which takes precedence, the result will stay the same. 
or at least from what I tested trying to come up with something that would break it and couldn't.
I'd say they can coexist without any issue, you can have both a blacklist and whitelist at the same time, and you are the one responsible for making it work the way you like to autogen your changeling, but these are simply options you have to work with.
so Im happy to apply necessary changes if you can give me an example that this might break or not work as it should

ljharb replied:
I think the order matters if there's a file that's covered by both patterns.

rouzwelt replied:
well maybe add a new flag so the precedence can be chosen?

ljharb replied:
is there a use case to provide both? perhaps the question could be deferred by making the two flags mutually exclusive.

rouzwelt replied:
well I assume there can be a case that both can be used, but probably that would be less than 5% of cases. but if you think making thme exclusive is a safer option, we can go with that

ljharb replied:
It can always be allowed later without being semver-major, but it can't be disallowed later without being semver-major.

ljharb commented on README.md:
```diff
@@ -21,40 +21,42 @@ Usage: auto-changelog [options]
-      --issue-pattern [regex]         # override regex pattern for issues in commit messages
-      --breaking-pattern [regex]      # regex pattern for breaking change commits
-      --merge-pattern [regex]         # add custom regex pattern for merge commits
-      --ignore-commit-pattern [regex] # pattern to ignore when parsing commits
```
this makes it a breaking change.

rouzwelt replied:
Im sorry I didn't understand this one, what is the issue?

ljharb replied:
you're removing `--ignore-commit-pattern`

rouzwelt replied:
yea so it is replaced with `--exclude-commit-pattern`, isnt it?

rouzwelt replied:
just for consistency, `include` and `exclude`

ljharb replied:
Yes, but removing anything is a breaking change.

rouzwelt replied:
ah I see, so what do you suggest? going back with `-- ignore-commit-pattern` ?

ljharb replied:
Probably that's simplest, yes.

rouzwelt replied:
sure

rouzwelt commented on README.md:
```diff
@@ -21,40 +21,42 @@ Usage: auto-changelog [options]
+      --merge-pattern [regex]           # add custom regex pattern for merge commits
+      --ignore-commit-pattern [regex]   # pattern to ignore when parsing commits
+      --include-commit-pattern [regex]  # pattern to include when parsing commits, merge commits are be included by default unless "--strict-include" is checked
+      --strict-include                  # strictly include only commits that match the pattern by --include-commits-pattern
```
yea, so when you use the `include pattern`, if you dont put something in your regex to include merge commits and PRs then it will not include them, so what this does is, by default all merge commits are included when you use `include commit`, meaning all merge commits + all the other commits that match `include pattern`. however if you wan to make it strict, meaning strictly include by the provided regex, the. you'd be using that flag

ljharb replied:
so `--strict-include` is sugar for `--merge-pattern=''` or something?

rouzwelt replied:
so basically, having this lets you not have to include the merge commit regex in your `include pattern` regex if you want merge commits to be part of the included commits

rouzwelt replied:
> so `--strict-include` is sugar for `--merge-pattern=''` or something?

no, let me explain with an example, I have this as my `include pattern` `/^fix/`.
so what happens now is all the commits that match this will be included, but merge commits if they don't match this wont be included, however you most probably want all the merge commits (PRs) to be included always (with or without custom `mergeCommit` option), 
so by default the `include pattern` will include all the merge commits (with or without custom `mergeCommit` option). but in case you want the including be strict to your provided pattern, you will use `strict-include` flag

rouzwelt replied:
@ljharb maybe if you think it should be removed and provide an explaination for how to achive a `--strict-include` by doing `--merge-comiit=''` ?

ljharb replied:
That sounds simpler to me.

rouzwelt replied:
ok fair enough, I'll update it

rouzwelt replied:
@ljharb well actually that would be a sugar for `no merge commit pattern`, meaning there would be no difference between merge commits and regualr commits

ljharb replied:
I guess i'd assume that the strict thing always applies, and that `include pattern` is a filter for "all the commits", and then whatever's left is what has the normal logic applied?

rouzwelt replied:
@ljharb what we can probably do is, to make the `getMergePattern()` return `/[]/` in case the `--merge-pattern=''`

rouzwelt replied:
> I guess i'd assume that the strict thing always applies, and that `include pattern` is a filter for "all the commits", and then whatever's left is what has the normal logic applied?

well imo, PRs and merge commits have higher value, so usually you would want them included, but probably dont want to go through a crazy regex if you want to include certain commits + all merge commits

ljharb replied:
hmm. it seems like the use cases motivating this PR probably needs a bit more fleshing out :-)

it's helpful to think in terms of minimal modification to existing functionality. If it takes two conceptual changes, then it's nice to figure out a way to split them into two separately motivated PRs.

rouzwelt replied:
maybe a better practice and compromise is to have `--include-commit-pattern` strict by default, and then provide a seperate flag to '--include-merge-commits`

rouzwelt replied:
> hmm. it seems like the use cases motivating this PR probably needs a bit more fleshing out :-)
> 
> it's helpful to think in terms of minimal modification to existing functionality. If it takes two conceptual changes, then it's nice to figure out a way to split them into two separately motivated PRs.

valid, well let me know what do you think about my last suggestion, and if you'd still prefer to split into another PR for that or not

rouzwelt replied:
@ljharb thinking now, it would make more sense to split out, because the same situation can apply to `--ignore-commit-pattern` where you may want to ignore commits but still include all marge commits

rouzwelt commented:
hey man @ljharb , let me know if everything is good for this to get merged, I'll open the next PR as soon as this one gets merged. or if you'd prefer I can open the next PR to be merging into this one?

ljharb commented:
@rouzwelt i'm reviewing as an interested user, i'm not a maintainer here.

rouzwelt commented:
@cookpete Hey man, I guess you have already looked into this PR, so I was wondering what would be the next steps? Im hoping this can get merged if all looks ok from your side?

rouzwelt commented:
@cookpete hey man, I hope I can get a response from you regarding this PR, please let me know your thoughts about this, Im really hoping this Pr can be considered, it is really big feature that I need with this package

bombillazo commented:
Hey @cookpete , any chance to get a review for this PR? I'd also be interested in helping maintain this if necessary.

rouzwelt commented:
> Hey @cookpete , any chance to get a review for this PR? I'd also be interested in helping maintain this if necessary.

thanks man for the support, let's wait and see what @cookpete thinks about this, he seems busy :D
but Im hoping we can have a resolution for this as soon as we get a response from @cookpete

cookpete commented:
Thanks @rouzwelt for the PR and @ljharb for the review. 

I’ve simplified the logic and renamed the option to `--commit-pattern` to be consistent with the other `--*-pattern` options.

I’ve also allowed both `--commit-pattern` and `--ignore-commit-pattern` to be used, for the rare case that you want to include a certain commit pattern _and then_ exclude some of those commits.

Sorry for the delay! Will publish a new version shortly after I’ve skimmed through the other issues and PRs.

rouzwelt commented:
> Thanks @rouzwelt for the PR and @ljharb for the review. 
> 
> I’ve simplified the logic and renamed the option to `--commit-pattern` to be consistent with the other `--*-pattern` options.
> 
> I’ve also allowed both `--commit-pattern` and `--ignore-commit-pattern` to be used, for the rare case that you want to include a certain commit pattern _and then_ exclude some of those commits.
> 
> Sorry for the delay! Will publish a new version shortly after I’ve skimmed through the other issues and PRs.

awesome man, thank you so much 🙏🏻

bombillazo commented:
Epic! Thanks!

cookpete commented:
> Is there a reason why its not yet published?

Because I forgot to publish this when I said I would (over a year ago) and then life got very busy.

Published in `2.5.0` 🎉

bombillazo commented:
thanks !
